electrumsv/network.py

```diff
@@ -1027,305 +297,157 @@ async def _main_task(self) -> None:
-        self.sessions_changed_event.set()
-        self.sessions_changed_event.clear()
-        # Disconnect the old main session, if any, in order to lose scripthash
-        # subscriptions.
-        if old_main_session:
-            if reason == SwitchReason.user_set:
-                old_main_session.server.state.retry_delay = 0
-            await old_main_session.close()
-        self.trigger_callback(NetworkEventNames.GENERIC_STATUS)
-
-    def add_electrumx_server(self, server_key: SVServerKey) \
-            -> None:
-        """
-        Add a new electrumx server.
-        """
-        if server_key in SVServer.all_servers:
-            raise KeyError("server already exists")
-
-        if server_key.protocol not in 'st':
-            raise ValueError(f'unknown protocol: {server_key.protocol}')
-
-        # This will register the server and make it available to the server connection logic
-        # to make use of. The server will also be persisted when the network shuts down.
-        SVServer.unique(server_key.host, server_key.port, server_key.protocol)
+            try:
+                headers_client = await self.get_headers_client()
+                if not headers_client:
+                    logger.debug("There are no suitable ESV Reference Servers available. "
+                                 "Re-checking in %s seconds", WAIT_TIME)
+                    await asyncio.sleep(WAIT_TIME)
+                    continue
+
+                # Will not proceed past this point until initial headers sync completes
+                await self._synchronize_initial_headers_async()
+
+                new_tip: TipResponse
+                async for new_tip in headers_client.subscribe_to_headers():
+                    logger.debug("Got new tip: %s", new_tip)
+                    await self._on_new_chain_tip(new_tip)
+
+            # TODO(1.4.0) Add the server to a ban list with a timeout
+            except ServiceUnavailableError as e:
+                pass
 
-    async def update_electrumx_server_async(self, existing_key: SVServerKey,
-            updated_key: SVServerKey) -> None:
-        """
-        Update the connection parameters for a given server instance.
+    def notify_wallets_new_tip(self, new_tip: Header, new_chain: Chain) -> None:
+        # Some merkle proofs can arrive before the corresponding header. Therefore we need to
+        # notify the 'late headers' background worker of any new headers.
+        logger.info("Notifying wallets of the new tip hash: %s, height: %s",
+            hash_to_hex_str(new_tip.hash), new_tip.height)
+        for wallet in self._wallets:
+            message = (new_tip, new_chain)
+            wallet._late_header_worker_queue.put_nowait((PendingHeaderWorkKind.NEW_HEADER, message))
+
+    async def notify_wallets_reorg_async(self, orphaned_block_hashes: List[bytes]) -> None:
+        logger.info("Notifying wallets of the the reorg")
```

Comment:
This is not enough. We have discussed how this needs to fix the existing problems with reorgs. Namely, that a wallet that is not open when the reorg is detected, never knows about the reorg. There needs to be some handling on wallet startup that checks to see if it has been reorged as well, and that reorgs. This should be similar to the calling logic, but generalised and referring to the persisted `last_tip_hash` value.